pkg/acsengine/pki.go

```diff
@@ -52,61 +52,119 @@ func CreatePki(extraFQDNs []string, extraIPs []net.IP, clusterDomain string, caP
 		errors <- err
 	}()
 
 	go func() {
 		var err error
 		organization := make([]string, 1)
 		organization[0] = "system:masters"
-		clientCertificate, clientPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, nil, nil, organization)
+		clientCertificate, clientPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, false, nil, nil, organization)
 		errors <- err
 	}()
 
 	go func() {
 		var err error
 		organization := make([]string, 1)
 		organization[0] = "system:masters"
-		kubeConfigCertificate, kubeConfigPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, nil, nil, organization)
+		kubeConfigCertificate, kubeConfigPrivateKey, err = createCertificate("clientclient", caCertificate, caPrivateKey, false, false, nil, nil, organization)
 		errors <- err
 	}()
 
+	go func() {
+		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
+		ip := net.ParseIP("127.0.0.1").To4()
+		peerIPs := append(extraIPs, ip)
+		etcdServerCertificate, etcdServerPrivateKey, err = createCertificate("etcdserver", caCertificate, caPrivateKey, true, true, nil, peerIPs, organization)
+		errors <- err
+	}()
+
+	go func() {
+		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
+		ip := net.ParseIP("127.0.0.1").To4()
+		peerIPs := append(extraIPs, ip)
+		etcdClientCertificate, etcdClientPrivateKey, err = createCertificate("etcdclient", caCertificate, caPrivateKey, true, false, nil, peerIPs, organization)
+		errors <- err
+	}()
+
+	etcdPeerCertPairs = make([]*PkiKeyCertPair, masterCount)
+	for i := 0; i < masterCount; i++ {
+		go func(i int) {
+			var err error
+			organization := make([]string, 1)
+			organization[0] = "system:masters"
+			ip := net.ParseIP("127.0.0.1").To4()
+			peerIPs := append(extraIPs, ip)
+			etcdPeerCertificate := new(x509.Certificate)
+			etcdPeerPrivateKey := new(rsa.PrivateKey)
+			etcdPeerCertificate, etcdPeerPrivateKey, err = createCertificate("etcdpeer", caCertificate, caPrivateKey, true, false, nil, peerIPs, organization)
+			etcdPeerCertPairs[i] = &PkiKeyCertPair{CertificatePem: string(certificateToPem(etcdPeerCertificate.Raw)), PrivateKeyPem: string(privateKeyToPem(etcdPeerPrivateKey))}
+			errors <- err
+		}(i)
+	}
+
 	e1 := <-errors
 	e2 := <-errors
 	e3 := <-errors
+	e4 := <-errors
+	e5 := <-errors
+	e6 := <-errors
 	if e1 != nil {
-		return nil, nil, nil, e1
+		return nil, nil, nil, nil, nil, nil, e1
 	}
 	if e2 != nil {
-		return nil, nil, nil, e2
+		return nil, nil, nil, nil, nil, nil, e2
 	}
 	if e3 != nil {
-		return nil, nil, nil, e2
+		return nil, nil, nil, nil, nil, nil, e3
+	}
+	if e4 != nil {
+		return nil, nil, nil, nil, nil, nil, e4
+	}
+	if e5 != nil {
+		return nil, nil, nil, nil, nil, nil, e5
 	}
+	if e6 != nil {
+		return nil, nil, nil, nil, nil, nil, e6
+	}
+
+	time.Sleep(10 * time.Second)
```

Comment:
I searched and looks golang can use sync.WaitGroup. Like this: https://stackoverflow.com/questions/46906647/wait-result-of-multiple-goroutines.

Sorry I am still new to golang and I may be wrong here. Just FYI.

Reply:
I actually had to add 2 error variables since each go routine is a separate error. Checking all the errors at the end waits for all the routines to finish. Thanks for the suggestion!